proposals/infra/INF-0004-register-types-and-diagnostics.md

```diff
@@ -0,0 +1,520 @@
+to the decl. The attribute only appertains to HLSL buffer objects, or external
+global variables. `err_attribute_wrong_decl_type_str` will be emitted in this case.
+
+Next, determine if the decl is a `cbuffer` or `tbuffer`. These are special in that they have
+their own decl class type, `HLSLBufferDecl`. To determine if the Decl is a `HLSLBufferDecl`,
+the compiler dynamically casts the Decl to an `HLSLBufferDecl`, and then if it succeeds,
+checks isCBuffer(). If true, the Decl is a cbuffer, otherwise it is a tbuffer. In either case,
+we know the decl is a `Resource` and can infer the resource class (`cbv` or `srv`
+respectively). Otherwise, if the dynamic cast fails, we cast to `VarDecl`, and check 
+to see if the ResourceClassAttr attribute is present. If it is, the `Resource` flag can be 
+set, and the corresponding resource class flag will be set. If the `VarDecl` dynamic cast
+fails, then there is some critical validation error, since any Decl with the register 
+annotation should either be a `HLSLBufferDecl` or a `VarDecl`.
+
+If the decl has no resource attribute, then check if the decl has an arithmetic type,
+and if so, we know that the decl is a `Basic` type. 
+If the decl isn't any of those types, then we check if it is a record type.
+If so, the decl is a UDT, so the `UDT` flag is set. For a `UDT`, recurse through the 
+members of the UDT and its bases, collecting information on what types of resources 
+are contained by the UDT, setting corresponding class flags (`SRV`, `UAV`, `CBV`, or 
+`Sampler`). Also track whether there are any other numeric members and set the 
+`ContainsNumeric` flag. Also, apply any template expansions on the UDT, so that all
+types of members in the UDT are defined. Types of interest may be contained in arrays
+of arbitrary dimension, so UDT analysis also requires drilling into arrays to determine
+if the underlying element type is numeric, or any resource class type of interest.
+If the decl is not a UDT, then the `other` flag is set.
+
+Below is some pseudocode to describe the flag-setting process:
+
+```
+if the decl is groupshared:
+  set other flag
+
+if the Decl is not inside a cbuffer or tbuffer:
+  if the Decl type is a numeric type:
+    set DefaultGlobals
+
+if the Decl is an HLSLBufferDecl:
+  set resource
+  if cbuffer: set cbv
+  else: set srv
+else if the Decl is a VarDecl:
+  if resource attribute: 
+    get resource attribute, set resource and corresponding resource class flag
+  else:
+    if Decl is vector, matrix, or otherwise numeric type:
+      set basic flag
+    else if Decl is struct or class type (if it is a udt):
+      set udt flag
+      recurse through members, set resource class flags that are found
+      set ContainsNumeric if the UDT contains a numeric member
+    else:
+      set other flag
+else:
+  raise error: (unknown decl type)
+```
+
+
+### Diagnostic emisison
+
+Now enough information has been gathered to determine the right diagnostics to emit, if any.

```

Comment:
I don't think I'm going to be able to read this all carefully enough to verify that it is correct. I'm basing most of my review on the tables of diagnostics and the examples shown above. I suspect that the full details of _how_ these diagnostics are determined would be best described in code.

Reply:
(resolving: no action necessarily required)